File: problems/two-sum/two_sum.rb
# @param {Integer[]} nums
# @param {Integer} target
# @return {Integer[]}
def two_sum(numbers, target)
  index_of = {}
  numbers.each_with_index do |number, index|
    residual = target - number
    if index_of.key?(residual)
      return [index_of[residual], index]
    end
    index_of[number] = index
  end
  return [-1, -1]
end

File: problems/two-sum/two_sum.py
from collections import defaultdict


class Solution(object):
    def twoSum(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: List[int]
        """
        hash_map = defaultdict(lambda: -1)
        for i, x in enumerate(nums):
            # Assume there is no conflict numbers
            hash_map[x] = i

        for i, x in enumerate(nums):
            j = hash_map[target - x]
            if j != -1 and j != i:
                return [i + 1, j + 1]
        return list()

if __name__ == "__main__":
    nums = [2, 7, 11, 15]
    target = 10
    print Solution().twoSum(nums, target)
